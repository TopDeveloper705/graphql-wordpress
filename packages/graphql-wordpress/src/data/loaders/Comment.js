import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import Comment from 'data/Comment';

// Dataloader expects IDs that can be read by the REST API

export default function getCommentLoaders() {
  const endpoint = Comment.getEndpoint();

  const commentLoader = new Dataloader(ids =>
    fetchData(endpoint, {
      qs: { include: ids, orderby: 'include', per_page: 100 },
    }).then(({ data: { body } }) => body)
  );

  return {
    load: async id => {
      const data = await commentLoader.load(id);
      return data ? Object.assign(new Comment(), data) : null;
    },
  };
}
