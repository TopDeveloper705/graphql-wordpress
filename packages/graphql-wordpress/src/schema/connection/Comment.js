import { GraphQLID, GraphQLString } from 'graphql';
import {
  connectionArgs,
  connectionFromArraySlice,
  fromGlobalId,
  connectionDefinitions,
} from 'graphql-relay';
import Comment from 'data/Comment';
import CommentType from 'type/Comment';
import COMMENT_ORDERBY from 'enum/CommentOrderby';
import ORDER from 'enum/Order';

const { connectionType: CommentConnection } = connectionDefinitions({
  nodeType: CommentType,
});

export default {
  type: CommentConnection,
  args: {
    post: {
      type: GraphQLID,
      description: 'Limit result set to resources assigned to a specific post id.',
    },
    slug: {
      type: GraphQLString,
      description: 'Limit result set to resources assigned to a specific post slug.',
    },
    orderby: {
      type: COMMENT_ORDERBY,
      description: 'Sort collection by object attribute.',
    },
    order: {
      type: ORDER,
      description: 'Order sort attribute ascending or descending.',
    },
    ...connectionArgs,
  },
  description: 'A list of results',
  resolve: async (root, args, context, { rootValue: { loaders: { Post } } }) => {
    const connectionArguments = {};
    const params = Object.assign({}, args);
    if (params.first) {
      params.per_page = params.first;
      connectionArguments.first = params.first;
      params.order = 'asc';
    } else if (params.last) {
      params.per_page = params.last;
      connectionArguments.last = params.last;
    } else {
      params.per_page = 100;
    }

    if (params.post) {
      params.post = fromGlobalId(params.post).id;
    } else if (params.slug) {
      params.post = await Post.loadBySlug(params.slug).then(post => post.id);
      delete params.slug;
    }

    delete params.first;
    delete params.last;
    delete params.after;
    delete params.before;

    return Comment.collection(params).then(({ items, total }) =>
      connectionFromArraySlice(items, connectionArguments, {
        arrayLength: total ? parseInt(total, 10) : items.length,
        sliceStart: params.offset || 0,
      })
    );
  },
};
