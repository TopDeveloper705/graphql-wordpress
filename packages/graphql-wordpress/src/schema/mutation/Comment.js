import { GraphQLID, GraphQLNonNull, GraphQLString } from 'graphql';
import { mutationWithClientMutationId, fromGlobalId } from 'graphql-relay';
import Comment from 'data/Comment';
import CommentType from 'type/Comment';

const requiredFields = {
  authorEmail: {
    type: new GraphQLNonNull(GraphQLString),
  },
  authorName: {
    type: new GraphQLNonNull(GraphQLString),
  },
  content: {
    type: new GraphQLNonNull(GraphQLString),
  },
  post: {
    type: new GraphQLNonNull(GraphQLID),
  },
};

const createCommentMutation = (opts = {}) =>
  mutationWithClientMutationId(
    Object.assign(
      {},
      {
        name: '<FixMe>',
        inputFields: {
          ...requiredFields,
        },
        outputFields: {
          status: {
            type: GraphQLString,
          },
          comment: {
            type: CommentType,
          },
          cookies: {
            type: GraphQLString,
          },
        },
        mutateAndGetPayload: payload => ({
          comment: payload,
          status: '<FixMe>',
          cookies: '<FixMe>',
        }),
      },
      opts
    )
  );

export default {
  addComment: createCommentMutation({
    name: 'AddComment',
    inputFields: {
      ...requiredFields,
      authorUrl: {
        type: GraphQLString,
      },
      parent: {
        type: GraphQLID,
      },
    },
    mutateAndGetPayload: input => {
      if (!input.author && !(input.authorEmail && input.authorName)) {
        return Promise.reject('You must provide author data to create a comment.');
      }

      if (!input.post) {
        return Promise.reject('You must provide a post to assign the comment to.');
      }

      const form = Object.assign({}, input);
      form.post = fromGlobalId(input.post).id;
      if (input.parent) {
        form.parent = fromGlobalId(input.parent).id;
      }
      return Comment.create(form);
    },
  }),
  updateComment: createCommentMutation({
    name: 'UpdateComment',
    inputFields: {
      id: {
        type: new GraphQLNonNull(GraphQLID),
      },
      content: {
        type: new GraphQLNonNull(GraphQLString),
      },
      token: {
        type: new GraphQLNonNull(GraphQLString),
      },
    },
    mutateAndGetPayload: input => {
      if (!input.id) {
        return Promise.reject('You must specify a comment ID to update.');
      }

      const form = Object.assign({}, input);
      form.id = fromGlobalId(form.id).id;
      return Comment.update(form);
    },
  }),
  deleteComment: createCommentMutation({
    name: 'DeleteComment',
    inputFields: {
      id: {
        type: new GraphQLNonNull(GraphQLID),
      },
      token: {
        type: new GraphQLNonNull(GraphQLString),
      },
      post: {
        type: new GraphQLNonNull(GraphQLString),
      },
    },
    outputFields: {
      status: {
        type: GraphQLString,
      },
    },
    mutateAndGetPayload: input => {
      if (!input.id) {
        return Promise.reject('You must specify a comment ID to update.');
      }

      const form = Object.assign({}, input);
      form.id = fromGlobalId(form.id).id;
      form.post = fromGlobalId(form.post).id;
      return Comment.delete(form);
    },
  }),
};
