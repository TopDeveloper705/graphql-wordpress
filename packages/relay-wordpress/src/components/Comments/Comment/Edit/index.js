// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import { formField } from '@wonderboymusic/graphql-wordpress-components';
import {
  EditForm,
  SubmitButton,
  EditCancelButton,
} from '@wonderboymusic/graphql-wordpress-components/lib/Comments';
import UpdateCommentMutation from 'mutations/UpdateComment';
import type { EditCommentProps, EditCommentState } from 'relay-wordpress';

export default class Edit extends React.Component<EditCommentProps, EditCommentState> {
  static contextTypes = {
    relay: PropTypes.object,
  };

  constructor(props, context: { relay: Object }) {
    super(props, context);

    this.state = {
      content: props.comment.content.raw,
    };
  }

  onEdit = (e: Event) => {
    e.preventDefault();

    this.props.onEditSubmit();

    const variables = {
      input: {
        id: this.props.comment.id,
        content: this.state.content,
        token: this.props.token,
      },
    };

    UpdateCommentMutation.commit(this.context.relay.environment, variables, this.props.comment);
  };

  onChange = (e: Event & { target: HTMLInputElement }) => {
    this.setState({
      [e.target.name]: e.target.value,
    });
  };

  render() {
    return (
      <EditForm onSubmit={e => e.preventDefault()}>
        <textarea
          className={formField}
          rows="6"
          name="content"
          value={this.state.content}
          onChange={this.onChange}
        />
        <SubmitButton type="submit" onClick={this.onEdit}>
          Submit
        </SubmitButton>
        <EditCancelButton type="reset" onClick={this.props.onEditSubmit}>
          Cancel
        </EditCancelButton>
      </EditForm>
    );
  }
}
