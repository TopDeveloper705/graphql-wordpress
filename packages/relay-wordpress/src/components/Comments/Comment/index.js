// @flow
import * as React from 'react';
import md5 from 'md5';
import { graphql, createFragmentContainer } from 'react-relay';
import { withCookies } from 'react-cookie';
import {
  CommentWrapper,
  Meta,
  Image,
  Author,
  Time,
  Content,
  Reply,
  activeReply,
  Actions,
  EditButton,
  DeleteButton,
} from '@wonderboymusic/graphql-wordpress-components/lib/Comments';
import { injectIntl } from 'react-intl';
import type { intlShape } from 'react-intl';
import { AUTHOR_EMAIL_COOKIE } from 'components/Comments/constants';
import DeleteCommentMutation from 'mutations/DeleteComment';
import type { Comment as CommentType } from 'relay-wordpress';
import EditComment from './Edit';

type CommentProps = {
  cookies: any,
  active: boolean,
  setReplyTo: (id: string) => void,
  comment: CommentType,
  intl: intlShape,
  relay: Object,
};

type CommentState = {
  editing: boolean,
};

/* eslint-disable react/no-danger */
/* eslint-disable react/forbid-prop-types */

@withCookies
class Comment extends React.Component<CommentProps, CommentState> {
  editToken = null;

  state = {
    editing: false,
  };

  onClick = (id: string) => {
    if (this.props.active) {
      this.props.setReplyTo('');
    } else {
      this.props.setReplyTo(id);
    }
  };

  onEditClick = () => {
    this.setState({
      editing: true,
    });
  };

  onEditSubmit = () => {
    this.setState({
      editing: false,
    });
  };

  onDelete = () => {
    DeleteCommentMutation.commit(this.props.relay.environment, {
      comment: this.props.comment,
      token: this.editToken,
    });
  };

  viewerOwns() {
    const { comment, cookies } = this.props;
    const authorEmail = cookies.get(AUTHOR_EMAIL_COOKIE);
    if (!authorEmail) {
      return false;
    }
    const tokenKey = encodeURIComponent(`token_${comment.id}`);
    this.editToken = cookies.get(tokenKey);
    if (!this.editToken) {
      return false;
    }
    const values = md5(`${comment.id}${authorEmail}`);
    return values === this.props.comment.authorHash;
  }

  render() {
    const {
      comment: {
        id,
        date,
        authorUrl,
        authorName,
        authorAvatarUrls,
        content: { rendered: content },
      },
    } = this.props;
    const avatar = authorAvatarUrls && authorAvatarUrls.find(data => data.size === 48);
    let authorDisplay = authorName;
    if (authorUrl) {
      authorDisplay = <a href={authorUrl}>{authorName}</a>;
    }

    let commentContent = null;
    if (this.state.editing) {
      commentContent = (
        <EditComment
          token={this.editToken}
          comment={this.props.comment}
          onEditSubmit={this.onEditSubmit}
        />
      );
    } else {
      commentContent = <Content dangerouslySetInnerHTML={{ __html: content }} />;
    }

    return (
      <CommentWrapper>
        <Meta>
          {avatar ? <Image alt="" src={avatar.url} /> : null}
          <Author>{authorDisplay}</Author>
          <Time>{this.props.intl.formatRelative(date)}</Time>
        </Meta>
        {commentContent}
        <Reply className={this.props.active ? activeReply : null} onClick={() => this.onClick(id)}>
          â†µ
        </Reply>
        {this.viewerOwns() &&
          !this.state.editing && (
            <Actions>
              <EditButton onClick={this.onEditClick}>Edit</EditButton>
              <DeleteButton onClick={this.onDelete}>Delete</DeleteButton>
            </Actions>
          )}
      </CommentWrapper>
    );
  }
}

export default createFragmentContainer(
  injectIntl(Comment),
  graphql`
    fragment Comment_comment on Comment {
      id
      authorName
      authorUrl
      authorHash
      date
      content {
        rendered
        raw
      }
      authorAvatarUrls {
        size
        url
      }
      parent
      post
    }
  `
);
