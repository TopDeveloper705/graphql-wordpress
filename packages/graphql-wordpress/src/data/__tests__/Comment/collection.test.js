import Comment from 'data/Comment';

const backupEnv = Object.assign({}, process.env);

afterAll(() => {
  process.env = backupEnv;
});

jest.mock('../../utils', () =>
  jest.fn(() => ({
    data: {
      body: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }, { id: 5 }],
      headers: {
        'x-wp-total': 5,
      },
    },
  }))
);

describe('Test Comment data access', () => {
  test('Get endpoint', () => {
    expect(Comment.getEndpoint()).toMatchSnapshot();
  });

  test('Load Comment collection', async () => {
    const collection = await Comment.collection();
    expect(collection).toMatchSnapshot();
  });

  test('Get backup endpoint', () => {
    delete process.env.WP_COMMENTS_ENDPOINT;
    expect(Comment.getEndpoint()).toMatchSnapshot();
  });
});
