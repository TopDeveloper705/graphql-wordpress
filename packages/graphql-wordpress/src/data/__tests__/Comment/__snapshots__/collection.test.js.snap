// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Comment data access Get backup endpoint 1`] = `"graphql/v1/comments"`;

exports[`Test Comment data access Get endpoint 1`] = `"graphql/v1/comments"`;

exports[`Test Comment data access Load Comment collection 1`] = `
Object {
  "items": Array [
    Comment {
      "id": 1,
    },
    Comment {
      "id": 2,
    },
    Comment {
      "id": 3,
    },
    Comment {
      "id": 4,
    },
    Comment {
      "id": 5,
    },
  ],
  "total": 5,
}
`;
